lib/lists.nix

```diff
@@ -582,6 +582,38 @@ rec {
     # Input list
     list: sublist count (length list) list;
 
+  /* Return a list of all elements until a predicate matches an element.
+
+     Type: takeWhile :: (a -> Bool) -> [a] -> [a]
+
+     Example:
+       takeWhile (x: x < 3) [ 1 2 3 4 5 ]
+       => [ 1 2 ]
+       takeWhile (x: elem x ["a" "b"]) [ "a" "a" "b" "c" "a" ]
+       => [ "a" "a" "b" ]
+  */
+  takeWhile = pred: list:
+    if list == [] || !(pred (head list))
+      then []
+      else [ (head list) ] ++ (takeWhile pred (tail list));
```

Comment:
> nix lists are sized eagerly

They're implemented as arrays, so tail is O(n).

Reply:
@roberth I'm pretty @pennae meant that the entire function becomes $\mathcal{O}(n²)$ :slightly_smiling_face:

Reply:
My intent was only to clarify the quoted phrase, not to disagree.

Reply:
Assuming `builtins.genlist` is $\mathcal{O}(n)$ and `builtins.elemAt` $\mathcal{O}(1)$, I believe the newly pushed implementation should be $\mathcal{O}(n)$.

As for whether they are needed, I think that these are useful when working with modules. While not *as* useful as they are with streams, and maybe not as useful as the other functions present either, I think they express the programmers intent very clearly when working with things like port ranges or parsing config for assertions.

Reply:
> Assuming `builtins.genlist` is $\mathcal{O}(n)$ and `builtins.elemAt` $\mathcal{O}(1)$, I believe the newly pushed implementation should be $\mathcal{O}(n)$.

Seems correct.



> I think they express the programmers intent very clearly when working with things like port ranges

How would you use this with port ranges? If you're going to `takeWhile (x: x < n)`, wouldn't `filter` be more robust? In the module system, we generally assume that list items can appear in any order, because we don't want to depend on the ordering that follows from the actually rather awkward `imports` resolution.

> parsing config

I don't think they're suitable for parsing anything non-trivial, because the _n_ in O(_n_) is the size of the (large) returned value instead of the characters dropped. This makes repeated `dropWhile` perform in O(n²).